test/test_type_promotion.py

```diff
@@ -167,6 +167,14 @@ def test_half(self, device):
         default_tensor = torch.tensor(100000.0, device=device)
         self.assertEqual((half + default_tensor).dtype, torch.get_default_dtype())
 
+    @float_double_default_dtype
+    def test_bfloat16(self, device):
```

Comment:
This test is a good start. I think it needs a tweak, though:

- test that a bfloat16 tensor + an int, float, and complex number results in the correct dtype and correct value (testing for a real value would be nice)
- test that an int, float, and complex number + a bfloat16 tensor results in the correct dtype


It seems like we should have (do we have?) a separate test that takes tensors of all different dtypes (such a test can be constructed using the `@dtypes` decorator and passing in a product of all the dtypes, in which case the "dtype" will actually be one of the tuples you construct) and verifies that adding those tensors produces a tensor of the type indicated by promote types? .

Reply:
I actually agree that we should make a separate PR to test tensors of all different dtypes and only leave bfloat16-specific tests here (probably after this PR for cleanliness though). Looks like this file doesn't have such kind of test. I opened a new issue #41842 for this.

Also I think the verification of the correctness of results of addition should probably be done in the tests on the add operator, not a type promotion test here...

What do you think?

Reply:
The add operator in this case is just a surrogate for all ops that participate in type promotion. It's not perfect, and in the future we'll probably have more extensive tests that every type promotion operator works as expected, but it's a start.

Reply:
Still need to test all the cases discussed. Also, if this won't add the full type promotion vs. promote types test, then we should at least verify that bfloat16 + other tensor results in the correct type and a runtime error where appropriate.

What does promote_types(bfloat16 float16) return, by the way? None or a RuntimeError? That'd be good to validate, too.

Reply:
OK, I added some type promotion tests.

`promote_types(bfloat16 float16)` raises RuntimeError.

Reply:
Great, thanks for the info on promote_types. Last nit from me, then, would to be validate that promote_types(a, b) raises when a + b raises and returns the correct dtype otherwise. You can just add the following 4 lines (extra lines included for context)

```
with self.assertRaises(RuntimeError)
  torch.promote_types(bf, t)
with self.assertRaises(RuntimeError):
  bf + t
```

```
selft.assertEqual(torch.promote_types(bf, t), dtype)
self.assertEqual((bf + t).dtype, dtype)
```

```
selft.assertEqual(torch.promote_types(bf, t), torch.bfloat16)
self.assertEqual((bf + t).dtype, torch.bfloat16)
```

Then I think I'm satisfied. @nairbv, did you want to add a review, too?

Reply:
OK, fixed

Reply:
Thanks! Let's give @nairbv a chance to take a look.